src/uu/cksum/src/cksum.rs

```diff
@@ -5,109 +5,140 @@
-    let mut crc = (input as u32) << 24;
-
-    let mut i = 0;
-    while i < 8 {
-        let if_condition = crc & 0x8000_0000;
-        let if_body = (crc << 1) ^ 0x04c1_1db7;
-        let else_body = crc << 1;
-
-        // NOTE: i feel like this is easier to understand than emulating an if statement in bitwise
-        //       ops
-        let condition_table = [else_body, if_body];
-
-        crc = condition_table[(if_condition != 0) as usize];
-        i += 1;
+fn detect_algo(program: &str) -> (&'static str, Box<dyn Digest + 'static>, usize) {
+    match program {
+        "sysv" => ("SYSV", Box::new(SYSV::new()) as Box<dyn Digest>, 512),
+        "bsd" => ("BSD", Box::new(BSD::new()) as Box<dyn Digest>, 1024),
+        "crc" => ("CRC", Box::new(CRC::new()) as Box<dyn Digest>, 256),
+        "md5" => ("MD5", Box::new(Md5::new()) as Box<dyn Digest>, 128),
+        "sha1" => ("SHA1", Box::new(Sha1::new()) as Box<dyn Digest>, 160),
+        "sha224" => ("SHA224", Box::new(Sha224::new()) as Box<dyn Digest>, 224),
+        "sha256" => ("SHA256", Box::new(Sha256::new()) as Box<dyn Digest>, 256),
+        "sha384" => ("SHA384", Box::new(Sha384::new()) as Box<dyn Digest>, 384),
+        "sha512" => ("SHA512", Box::new(Sha512::new()) as Box<dyn Digest>, 512),
+        "blake2b" => (
+            "BLAKE2",
+            Box::new(blake2b_simd::State::new()) as Box<dyn Digest>,
+            512,
+        ),
+        "sm3" => ("SM3", Box::new(sm3::Sm3::new()) as Box<dyn Digest>, 512),
+        _ => panic!("unknown algorithm"),
     }
-
-    crc
 }
 
-#[inline]
-fn crc_update(crc: u32, input: u8) -> u32 {
-    (crc << 8) ^ CRC_TABLE[((crc >> 24) as usize ^ input as usize) & 0xFF]
+struct Options {
+    algo_name: String,
+    digest: Box<dyn Digest + 'static>,
+    output_bits: usize,
 }
 
-#[inline]
-fn crc_final(mut crc: u32, mut length: usize) -> u32 {
-    while length != 0 {
-        crc = crc_update(crc, length as u8);
-        length >>= 8;
+#[allow(clippy::cognitive_complexity)]
+fn cksum<'a, I>(mut options: Options, files: I) -> UResult<()>
+where
+    I: Iterator<Item = &'a OsStr>,
+{
+    for filename in files {
+        let filename = Path::new(filename);
+        let stdin_buf;
+        let file_buf;
+        let not_file = filename == OsStr::new("-");
+        let mut file = BufReader::new(if not_file {
+            stdin_buf = stdin();
+            Box::new(stdin_buf) as Box<dyn Read>
+        } else if filename.is_dir() {
+            Box::new(BufReader::new(io::empty())) as Box<dyn Read>
+        } else {
+            file_buf =
+                File::open(filename).map_err_context(|| filename.to_str().unwrap().to_string())?;
+            Box::new(file_buf) as Box<dyn Read>
+        });
+        let (sum, sz) = digest_read(&mut options.digest, &mut file, options.output_bits)
+            .map_err_context(|| "failed to read input".to_string())?;
+
+        if options.algo_name == "SYSV" {
+            let width = 1;
```

Comment:
could you please document why this is special ?
same for bsd

Reply:
ping ? :)

Reply:
I was followed the implementation in `sum` actually.